Use offset tokens to ensure at-least-once semantics

istreeter opened the pull request:
Jira ref: [PDP-1975](https://snplow.atlassian.net/browse/PDP-1975)

Until now, we have (effectively) used response codes from Snowflake to tell us that the events are safely in Snowflake. This was done via the snowflake java sdk; specifically, if `.flush()` returns with no error then we assumed the events are in snowflake.

But we hit edge cases where snowflake would response with a successful response code, but the events were not fully ingested into the table. This edge case can happen around the time of schema evolution.

After this PR, we use offset tokens to check the ingestion of a batch of events. We do not assume the batch is ingested to the table until `channel.getLatestCommittedOffsetToken()` returns the expected offset.

[PDP-1975]: https://snplow.atlassian.net/browse/PDP-1975?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

snowplowcla commented:
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://github.com/snowplow/snowplow/wiki/CLA to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

nov1n commented:
I signed it.

snowplowcla commented:
Confirmed! @nov1n  has signed the Individual Contributor License Agreement. Thanks so much

rzats commented on transformer/src/main/scala/com/snowplowanalytics/snowflake/transformer/TransformerJob.scala:
```diff
@@ -52,13 +52,31 @@ object TransformerJob {
-      .map { e => Transformer.jsonify(e) }
+      .wholeTextFiles(jobConfig.input)
+      .flatMap{ case (fileName, fileContents) => {
+        fileContents.lines.zipWithIndex.map{case (event, index) =>
```
`zipWithIndex` isn't a great idea - it loads an entire file into a single node's memory, which is a no-no when working with Spark. As far as I can tell the only reason to do this is to have slightly more verbose logging output, so I'd not use this function at all.

rzats commented on transformer/src/main/scala/com/snowplowanalytics/snowflake/transformer/TransformerJob.scala:
```diff
@@ -13,15 +13,15 @@
 package com.snowplowanalytics.snowflake.transformer
 
 import org.apache.spark.sql.{SaveMode, SparkSession}
-
```
Unnecessary whitespace edits here.

nov1n commented:
> Hi @nov1n! I definitely like the idea! I think right now Snowflake Loader is overly optimistic about its input (although it never was a problem in our experience).
> 
> However, I think Spark provides more optimal solutions for this kind of dataset splitting, e.g. you can get some inspiration from RDB Shredder: https://github.com/snowplow/snowplow-rdb-loader/blob/master/shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala#L405-L422
> 
> Also please pay attention that our 0.5.0 version [will be using](https://github.com/snowplow-incubator/snowplow-snowflake-loader/pull/74) Scala Analytics SDK 0.4.0, which has much better API.
> 
> Thank you very much for raising this!

Thanks for the quick feedback. The problem addressed by this PR however is very pressing for us, and we're looking for a solution. Will you merge and release this in the near future if I address the feedback, or should I run a custom binary to get this functionality?

nov1n commented:
@chuwy could you get back to my question?

chuwy commented:
Hi @nov1n! Sorry, missed your last question. Couple of points here:

1. We're definitely interested in implementing bad rows for Snowflake Loader and we'll merge this feature if it will match criterions outlined below
2. No performance penalty. Its quite easy to run into skewing issues if you try to split streams (e.g. on two nodes cluster, one node writes 99% of good data and another 1% of bad data, so we're wasting 50% of cluster). If it is not possible to avoid that, we should at least make it optional (I don't think we ever stumbled upon invalid enriched data in transformer)
3. It should be based on top of ongoing 0.5.0 branch https://github.com/snowplow-incubator/snowplow-snowflake-loader/pull/74 as we changed several underlying libs significantly

Thanks!

nov1n commented on core/src/main/scala/com.snowplowanalytics.snowflake.core/Config.scala:
```diff
@@ -323,6 +324,21 @@ object Config {
     )
   )
 
+  object optionalS3FolderSerializer extends CustomSerializer[Option[S3Folder]]((formats) =>
```
I thought the same, but if I remove it I get a runtime exception [here](https://github.com/snowplow-incubator/snowplow-snowflake-loader/pull/79/files#diff-d9163363ee232782ef2656a544f5d41dR29): 
```
Exception in thread "main" java.lang.ClassCastException: java.lang.String cannot be cast to com.snowplowanalytics.snowflake.core.Config$S3Folder
	at com.snowplowanalytics.snowflake.transformer.TransformerJobConfig.errorOutput(TransformerJobConfig.scala:30)
```
It seems that at runtime it is a `String`, even though the compiler thinks it will be an `S3Folder`. Am I missing something?

chuwy commented:
Closing as this has been implemented as part of https://github.com/snowplow-incubator/snowplow-snowflake-loader/issues/85

istreeter committed:
Use offset tokens to ensure at-least-once semantics

Until now, we have (effectively) used response codes from Snowflake to
tell us that the events are safely in Snowflake. This was done via the
snowflake java sdk; specifically, if `.flush()` returns with no error
then we assumed the events are in snowflake.

But we hit edge cases where snowflake would response with a successful
response code, but the events were not fully ingested into the table.
This edge case can happen around the time of schema evolution.

After this commit, we use offset tokens to check the ingestion of a
batch of events. We do not assume the batch is ingested to the table
until `channel.getLatestCommittedOffsetToken()` returns the expected
offset.

istreeter committed:
Amendment: Bump default upload parallelism factor to 3.5

spenes commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow.snowflake/processing/Channel.scala:
```diff
@@ -166,6 +167,18 @@ object Channel {
+    channel: SnowflakeStreamingIngestChannel,
+    offsetToken: String
+  ): F[Unit] =
+    Sync[F].untilDefinedM {
```
I guess we don't need to limit the retries since we always expect it to be successful eventually, right ?

istreeter replied:
Yeah exactly.

You are right to be nervous about potentially infinite loops!  But here I think it is ok because:

1. We expect it to be successful eventually
2. In the worst case scenario where it never becomes successful (I don't know why this would ever happen) then eventually the health probe will become unhealthy because the source is not making progress, and the app will get restarted.
3. If an exception happens then it bubbles up, which breaks out of the loop

spenes reviewed (approved):
LGTM, very clean solution!
